### PR TITLE
Add support for getting and putting Lifecycle policies on s3 buckets

### PR DESCRIPTION
**Pull request by edhollandAL**
@motobob @kkuzmin plz review

### CODE REVIEW

**Comment by edhollandAL** [comment]
nb: i added a parser/encoder to convert to and from the XML policy document returned by AWS to simplify implementation in consuming s/w

File: src/erlcloud_s3.erl [hunk 1]
```diff
@@ -474,6 +508,92 @@ get_bucket_attribute(BucketName, AttributeName, Config)
...
+    case List of
+        [{_, _} | _] ->
+            lists:map(fun({Key, Value}) ->
+                              {key_to_name(Key), [to_string(Value)]}
+                      end, List);
+        Value ->
+            [to_string(Value)]
+    end.
+
+to_string(X) when is_list(X) ->
```
Comment (motobob) [inline src/erlcloud_s3.erl]: i really do not want to see these type of funs anywhere except _utils module

File: src/erlcloud_s3.erl [hunk 2]
```diff
@@ -34,6 +36,10 @@
          get_bucket_and_key/1
         ]).
 
+-ifdef(TEST).
```
Comment (motobob) [inline src/erlcloud_s3.erl:40]: do we really need to export it at all?

File: src/erlcloud_s3.erl [hunk 3]
```diff
@@ -363,12 +369,40 @@ put_bucket_policy(BucketName, Policy, Config)
...
+            Error ->
+                Error
+        end.
+
+-spec put_bucket_lifecycle(string(), binary()) -> ok.
+put_bucket_lifecycle(BucketName, Policy) ->
+    put_bucket_lifecycle(BucketName, Policy, default_config()).
+
+-spec put_bucket_lifecycle(string(), list() | binary(), aws_config()) -> ok.
+put_bucket_lifecycle(BucketName, Policy, Config)
```
Comment (motobob) [inline src/erlcloud_s3.erl:394]: make this fun use the one below as passthrow

File: src/erlcloud_s3.erl [hunk 4]
```diff
@@ -363,12 +369,40 @@ put_bucket_policy(BucketName, Policy, Config)
   when is_list(BucketName), is_binary(Policy), is_record(Config, aws_config) ->
     s3_simple_request(Config, put, BucketName, "/", "policy", [], Policy, []).
 
+-spec(get_bucket_lifecycle/1 :: (BucketName::string()) -> ok | {error, Reason::term()}).
```
Comment (kkuzmin) [inline src/erlcloud_s3.erl:375]: Please fix the spec of the return. It should be the same as `get_bucket_lifecycle/2`

File: src/erlcloud_s3.erl (hunk 2 above)
Comment (edhollandAL) [inline src/erlcloud_s3.erl:40]: Its exported so we can test it within the framework of the existing tests. Personally i'd prefer to have the eunit tests in the same file, but thats not how the rest of the library does it.

File: src/erlcloud_s3.erl (hunk 4 above)
Comment (edhollandAL) [inline src/erlcloud_s3.erl:375]: fixed

**Pull request merged**